### PR TITLE
[LLVMGPU] Improve how we distribute small inner shapes

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Improve the default kernel configuration for cases where the inner dimensions are smaller than the workgroup size we want to target. Also fix a bug in tile and distribute for loops that are not distributed.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 1]
```diff
@@ -462,6 +462,25 @@ static LogicalResult setRootDefaultConfig(func::FuncOp entryPoint,
         vectorSize = 1;
         break;
       }
+      // If the inner dimension is too small to have one element per thread
+      // reduce the workgroup size try to distribute amonst more dimensions.
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: typo: amongst

**Review by antiagainst — APPROVED** [review]
LGTM!

File: compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp [hunk 2]
```diff
@@ -104,6 +104,14 @@ static LogicalResult getTileAndDistributeConfig(
     tileSizes[loopId] = 0;
   }
 
+  // Loops with a tile size of 0 shouldn't be distributed.
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp]: AFAIR this should already be happening. Can you provide example of what is hitting this and add a test maybe?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp]: The test I added hits this case

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp]: If the tile size is zero, it shouldnt even be tiled... so maybe the issue is somewhere else?

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp]: Correct it is not tiled, the problem is that when we calculate the group size we will set the group dim size to 1 instead if skipping the distribution:
https://github.com/openxla/iree/blob/289b9a1ebf954be3995cc1895254d09f2623341f/compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp#L177

therefore we will have a mismatch in ids and number of groups

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp]: add a test to show the problem. Please take another look

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp]: IIUC, if you add
```
if (tileSizes[partitionedLoop] == 0) continue;
```
in the loop  here
https://github.com/openxla/iree/blob/289b9a1ebf954be3995cc1895254d09f2623341f/compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp#L205 it should also fix?

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp]: yes that works, changed it.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**